### PR TITLE
[14.0][ADD] hr_expense_from_purchase_request

### PR DESCRIPTION
**Pull request by kittiu**
This module allow using Purchase Request as helper to create expenses in expense report.

In expense report (hr.expense.sheet), user can now choose his/her existing purchase request
to help fill in multiple expense lines.

- [x] Test
- [x] Bug fix https://github.com/OCA/hr-expense/pull/43

### CODE REVIEW

**Comment by kittiu** [comment]
cc @newtratip

File: hr_expense_from_purchase_request/views/hr_expense_views.xml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+                <page
+                    name="prlines"
+                    string="Expense from PR"
+                    attrs="{'invisible': [('pr_line_ids', '=', [])]}"
+                >
+                    <p class="oe_grey">
+                        By click "Save", expenses based on purchase request lines, will be created into "Expense" tab.
+                    </p>
+                    <field name="pr_line_ids">
+                        <tree editable="bottom" create="0">
```
Comment (newtratip) [inline hr_expense_from_purchase_request/views/hr_expense_views.xml]: I think user should not delete `Expense from PR`, can edit only.

```suggestion
                        <tree editable="bottom" create="0" delete="0">
```

File: hr_expense_from_purchase_request/views/hr_expense_views.xml [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+                            <field name="name" force_save="1" readonly="1" />
+                            <field name="product_id" />
```
Comment (newtratip) [inline hr_expense_from_purchase_request/views/hr_expense_views.xml:32]: Change product poisiton to before description

File: hr_expense_from_purchase_request/views/hr_expense_views.xml [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+                            <field name="quantity" force_save="1" readonly="1" />
+                            <field name="product_uom_id" />
```
Comment (newtratip) [inline hr_expense_from_purchase_request/views/hr_expense_views.xml]: UOM show only group `Manage Multiple Units of Measure`

File: hr_expense_from_purchase_request/views/hr_expense_views.xml [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+                            <field name="unit_amount" force_save="1" readonly="1" />
+                            <field name="total_amount" force_save="1" readonly="1" />
+                            <field
+                                name="company_id"
```
Comment (newtratip) [inline hr_expense_from_purchase_request/views/hr_expense_views.xml]: groups="base.group_multi_company"

File: hr_expense_from_purchase_request/views/hr_expense_views.xml [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+                                force_save="1"
+                                readonly="1"
+                                optional="hide"
+                            />
+                            <field
+                                name="analytic_account_id"
```
Comment (newtratip) [inline hr_expense_from_purchase_request/views/hr_expense_views.xml]: groups="analytic.group_analytic_accounting"

File: hr_expense_from_purchase_request/views/hr_expense_views.xml [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                                force_save="1"
+                                readonly="1"
+                                optional="hide"
+                            />
+                            <field
+                                name="analytic_tag_ids"
```
Comment (newtratip) [inline hr_expense_from_purchase_request/views/hr_expense_views.xml]: groups="analytic.group_analytic_tags"

File: hr_expense_from_purchase_request/models/hr_expense.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+        comodel_name="product.product",
+        string="Product",
+        required=True,
+    )
+    product_uom_id = fields.Many2one(
+        comodel_name="uom.uom",
+        string="Unit of Measure",
+        required=True,
+    )
+    unit_amount = fields.Float(
```
Comment (newtratip) [inline hr_expense_from_purchase_request/models/hr_expense.py]: Monetary field

File: hr_expense_from_purchase_request/views/hr_expense_views.xml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+    <record id="view_hr_expense_sheet_form" model="ir.ui.view">
+        <field name="name">hr.expense.sheet.form.inherit</field>
+        <field name="model">hr.expense.sheet</field>
+        <field name="inherit_id" ref="hr_expense.view_hr_expense_sheet_form" />
+        <field name="arch" type="xml">
+            <field name="company_id" position="after">
+                <field name="employee_user_id" invisible="1" />
+                <field
+                    name="purchase_request_id"
+                    domain="[('requested_by', '=', employee_user_id)]"
```
Comment (newtratip) [inline hr_expense_from_purchase_request/views/hr_expense_views.xml]: groups="purchase_request.group_purchase_request_user"

**Review by newtratip — COMMENTED** [review]
My comment

**Comment by kittiu** [comment]
@newtratip thanks for review. All ffixed.

File: hr_expense_from_purchase_request/security/ir.model.access.csv [hunk 9]
```diff
@@ -0,0 +1,2 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+access_hr_expense_sheet_prline,access_hr_expense_sheet_prline,model_hr_expense_sheet_prline,base.group_user,1,1,1,1
```
Comment (newtratip) [inline hr_expense_from_purchase_request/security/ir.model.access.csv:2]:
```suggestion
access_hr_expense_sheet_prline,access_hr_expense_sheet_prline,model_hr_expense_sheet_prline,purchase_request.group_purchase_request_user,1,1,1,1
```

File: hr_expense_from_purchase_request/models/hr_expense.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        unit_amount = (
+            line.estimated_cost / line.product_qty if line.product_qty > 0 else 0
+        )
+        return {
+            "name": line.name,
+            "product_id": line.product_id.id,
+            "product_uom_id": line.product_uom_id.id,
+            "unit_amount": unit_amount,
+            "quantity": line.product_qty,
+            "total_amount": unit_amount * line.product_qty,
```
Comment (newtratip) [inline hr_expense_from_purchase_request/models/hr_expense.py]: `total_amount` should be computed field, for my extension module `unit_amount` no readonly field. If I changed this one `total_amount` is not updated.

**Review by newtratip — COMMENTED** [review]
Please send employee field to `hr.expense`

Now,
PR, request by `A`
Expense Report, select employee `A`

The user created `EXPENSE REPORT` by `B`
Employee in `EXPENSE` will be `B`

**Comment by kittiu** [comment]
@newtratip all fixed.

**Review by newtratip — APPROVED** [review]
LGTM :+1: Please add test script too.

File: hr_expense_from_purchase_request/models/hr_expense.py [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+            "reference": line.specifications,
+        }
+
+    @api.model
+    def create(self, vals):
+        sheet = super().create(vals)
+        if "purchase_request_id" in vals:
+            sheet.mapped("expense_line_ids").unlink()
+        sheet._do_process_from_purchase_request()
+        sheet.pr_line_ids.unlink()  # clean after use
```
Comment (newtratip) [inline hr_expense_from_purchase_request/models/hr_expense.py:72]: Make sure create expense when Purchase Request != False

Please change to:
```
if vals.get("purchase_request_id"):
     sheet.mapped("expense_line_ids").unlink()
     sheet._do_process_from_purchase_request()
     sheet.pr_line_ids.unlink() 
```

File: hr_expense_from_purchase_request/models/hr_expense.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+        return sheet
+
+    def write(self, vals):
+        res = super().write(vals)
+        if "purchase_request_id" in vals:
+            self.mapped("expense_line_ids").unlink()
+        self._do_process_from_purchase_request()
+        self.mapped("pr_line_ids").unlink()  # clean after use
```
Comment (newtratip) [inline hr_expense_from_purchase_request/models/hr_expense.py:80]: Make sure create expense when Purchase Request != False

Please change to:
```
if vals.get("purchase_request_id"):
     self.mapped("expense_line_ids").unlink()
     self._do_process_from_purchase_request()
     self.mapped("pr_line_ids").unlink()
```

**Review by newtratip — COMMENTED** [review]
I found something wrong in the expense report

1. Create new expense report with select expense line `(no selected purchase request)` and save it, expense line will be blank
2. Create new expense report with select purchase request and select `use pr for = Advance` and save it then click submit to manager, the adv sheet will create new exp line

File: hr_expense_from_purchase_request/models/hr_expense.py (hunk 12 above)
Comment (kittiu) [reply hr_expense_from_purchase_request/models/hr_expense.py:80]: Are you sure? It was my intention to always do the job.

Comment (newtratip) [reply hr_expense_from_purchase_request/models/hr_expense.py:80]: Yes, I'm sure

This was example the error
1. No line when save expense report
Peek 2021-05-26 10-22
2. Select PR and use pr for = Advance
Peek 2021-05-26 10-29

Comment (kittiu) [reply hr_expense_from_purchase_request/models/hr_expense.py:86]: I still don't face your issue. I can still use EX as normal (no line deleted). I will keep your finding in mind, if I saw it I will fix it.

Comment (kittiu) [reply hr_expense_from_purchase_request/models/hr_expense.py:80]: Fixed, thanks for reporting.

**Comment by kittiu** [comment]
@newtratip test script added. Please review again thanks!

**Comment by kittiu** [comment]
@newtratip @Saran440

**Comment by kittiu** [comment]
No longer use.

**Pull request closed without merging**